rapid7/metasploit-framework
# Improve secrets dump persistence for kerberos and add keytab export functionality

Depends on database model changes: https://github.com/rapid7/metasploit-credential/pull/165

Adds support for:
- Updating secrets dump module to save real KrbEncKeys which can later be used for encrypting/decrypting Kerberos tickets
- Adds export functionality to the `modules/auxiliary/admin/kerberos/keytab.rb` module, which is useful for decrypting wireshark traffic, as well as integrating with other tools
- Persists Kerberos tickets as creds, instead of loot

```
msf6 auxiliary(scanner/winrm/winrm_login) > run 192.168.123.144  domain=adf3.local username=Administrator password=p4$$w0rd Winrmauth=kerberos domaincontrollerrhost=192.168.123.13 winrmrhostname=win10-dc3.adf3.local

[*] 192.168.123.13:88 - Received a valid TGT-Response
[+] 192.168.123.13:88 - Received a valid TGS-Response
[*] 192.168.123.13:88 - TGS MIT Credential Cache saved to /Users/adfoster/.msf4/loot/20221110132053_default_192.168.123.13_mit.kerberos.cca_926716.bin
[+] 192.168.123.13:88 - Received a valid delegation TGS-Response
[+] 192.168.123.13:88 - Received AP-REQ. Extracting session key...
[+] 192.168.123.144:5985 - Login Successful: adf3.local\Administrator:p4$$w0rd
[*] Command shell session 7 opened (192.168.123.1:55735 -> 192.168.123.144:5985) at 2022-11-10 13:20:54 +0000
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/winrm/winrm_login) > creds
Credentials
===========

host             origin           service            public         private                                                                                              realm       private_type  JtR Format
----             ------           -------            ------         -------                                                                                              -----       ------------  ----------
192.168.123.13   192.168.123.13   88/tcp (kerberos)  Administrator  aes256-cts-hmac-sha1-96:56c3bf6629871a4e4b8ec894f37489e823bbaecc2a0a4a5749731afa9d158e0 (TRUNCATED)  ADF3.LOCAL  Krb enc key   
192.168.123.13   192.168.123.13   88/tcp (kerberos)  Administrator  tgs:http/win10-dc3.adf3.local:ccache:BQQAAAAAAAEAAAABAAAACkFERjMuTE9DQUwAAAANQWRtaW5pc3 (TRUNCATED)  ADF3.LOCAL  Krb ticket    
192.168.123.13   192.168.123.13   88/tcp (kerberos)  Administrator  tgt:krbtgt/ADF3.LOCAL:ccache:BQQAAAAAAAEAAAABAAAACkFERjMuTE9DQUwAAAANQWRtaW5pc3RyYXRvcg (TRUNCATED)  ADF3.LOCAL  Krb ticket    
192.168.123.144  192.168.123.144  5985/tcp (http)    Administrator  p4$$w0rd                                                                                             adf3.local  Password      

msf6 auxiliary(scanner/winrm/winrm_login) > 
```

Out of scope: I'd argue that the user/pass registered as coming from the 5985/tcp (http) service is wrong, as it's really been verified by Kerberos

## Verification

- Verify the ldap_query, windows/smb/psexec, smb_login (https://github.com/rapid7/metasploit-framework/pull/17175) - similar to https://github.com/rapid7/metasploit-framework/pull/17079
- Verify pkinit works. Steps: https://github.com/rapid7/metasploit-framework/pull/17117
- Verify ticket forging works. Steps: https://github.com/rapid7/metasploit-framework/pull/17077
